documentation api (#48)
Cyberwatch/cyberwatch_api_toolbox: template.md

> # Cyberwatch API Toolbox documentation

Reviewer: Renommer le ficher en documentation.md

## Available methods

#### Ping

> Sends a GET request to `/api/v2/ping` to validate that the API is working.

Reviewer: Send ou Sends ? @fakenine tu en penses quoi ?

Reviewer 2: Send à mon avis, car on propose d'effectuer une action

Author: :+1:

###### Usage example and expected result:

```python
>>> CBWApi(URL, API_KEY, SECRET_KEY).ping()
INFO:root:OK
True
```

#### Servers

Sends a GET request to `/api/v2/servers` to retrieve the list of all servers.

###### Usage example and expected result:

```python
>>> CBWApi(URL, API_KEY, SECRET_KEY).servers()
[<cbw_api_toolbox.cbw_objects.cbw_server.CBWServer ...]
```

#### Server

> Sends a GET request to `/api/v2/servers/{SERVER_ID}` to retrieve the informations of SERVER_ID.

Reviewer: Information est indénombrable

##### Usage example and expected result:

```python
>>> CBWApi(URL, API_KEY, SECRET_KEY).server(SERVER_ID)
[<cbw_api_toolbox.cbw_objects.cbw_server.CBWServer]
```

#### Update server

> Sends a PATCH request  `/api/v2/servers/{SERVER_ID}` to update the informations of SERVER_ID.

Reviewer: Information est indénombrable

```python
>>> CBWApi(URL, API_KEY, SECRET_KEY).update_server(SERVER_ID, INFO)
>>>True
```

#### Delete server

Sends a DELETE request `/api/v2/servers/{SERVER_ID}` to delete SERVER_ID.

```python
>>> CBWApi(URL, API_KEY, SECRET_KEY).delete_server(SERVER_ID)
>>>True
```

#### Remote accesses

Sends a GET request `/api/v2/remote_accesses` to retrieve the list of all remote accesses.

```python
>>> CBWApi(URL, API_KEY, SECRET_KEY).remote_accesses()
[<cbw_api_toolbox.cbw_objects.cbw_remote_access.CBWRemoteAccess...]
```

#### Remote access

> Sends a GET request `/api/v2/remote_accesses/{REMOTE_ACCESS_ID}` to retrieve the informations of REMOTE_ACCESS_ID.

Reviewer: Information est indénombrable

```python
>>> CBWApi(URL, API_KEY, SECRET_KEY).remote_access(REMOTE_ACCESS_ID)
[<cbw_api_toolbox.cbw_objects.cbw_remote_access.CBWRemoteAccess]
```

> ### Create remote access

Reviewer: Mettre ####

Sends a POST request `/api/v2/remote_accesses` to create a remote access

```python
>>> CBWApi(URL, API_KEY, SECRET_KEY).create_remote_access(INFO)
INFO:root:remote access successfully created {ADDRESS}
True
```

> ### Update remote access

Reviewer: Mettre ####

> Sends a PATCH request `/api/v2/remote_accesses/{REMOTE_ACCESS_ID}` to update the informations of REMOTE_ACCESS_ID

Reviewer: Information est indénombrable

| id                        | String (hash) | Unique id of the computer         | 4a78524087574c12453dea248a91cadb                  |
| agent_version             | String        | Agent version used for supervision in case of agent supervision. Empty for agentless connections          | DESKTOP-AAA111                                    |
| applications              | List          | List of the computer's applications |                                                   |
| boot_at                   | Date          | Date of the computer boot         | 2016-05-30T09:54:58.000+02:00                     |
| category                  | String        | Computer category                 | server                                            |
| criticality               | String        | Computer criticality              | criticality_medium                                |
| cve_announcements         | List          | All available listings            | CVE-2019-3842                                     |
| cve_announcements_count   | Int           | Number of CVE                     | 3                                                 |    

> | deploying_period          | List          | The time of deployment            | <table><tr><td>name:</td><td>"test"</td></tr><tr><td>autoplanning:</td><td>false</td></tr><tr><td>autoreboot:</td><td>false</td></tr><tr><td>start_time:</td><td>"00:00"</td><tr><td>end_time:</td><td>"06:00"</td></tr><tr><td>next_occurence:</td><td>"2019-04-27T00:00:00.000+02:00"</td></tr></table>|

Reviewer: Ce n'est pas une list

| description               | String        | Description of the machine        | "machine of production"                           |
| groups                    | List          | Groups of the machine             | "Production, Development"                         |
| hostname                  | String        | Name of the machine               | "XXX.XXX.XXX.XXX"                                 |

> | ignoring_policy           | List          | Ignoring policy of the computer   | <table><tr><td>name:</td> <td>"test"</td></tr><tr><td>keyword:</td><td>"test"</td></tr><tr><td>version:</td><td>""</td></tr></table>|

Reviewer: Ce n'est pas une liste

Reviewer: Mettre Object dans le 2ème colonne, puis dans la dernière : [IgnoringPolicy Object](#ignoring-policy) (le bon lien vers la section qui parle de cette object)

| last_communication        | String        | Last communication of the machine | "2019-04-08T15:46:22.000+02:00"                   |

> | os                        | List          | Os of the machine                 | <table><tr><td>key:</td><td>"ubuntu_1804_64"</td></tr><tr><td>name:</td><td>"Ubuntu 18.04"</td></tr><tr><td>arch:</td><td>"x86_64"</td></tr><tr><td><eol:</td><td>"2023-04-26T02:00:00.000+02:00"</td></tr><tr><td>short_name:</td><td>"Ubuntu 18.04"</td></tr><tr><td>type:</td><td>"Os::Ubuntu"</td></tr><tr><td>hash_index:</td><td>"c3f4de66dd34cec0156bb439002bc4..."</td></tr><tr><td>created_at:</td><td>"2019-04-08T11:01:20.000+02:00"</td></tr><tr><td>updated_at:</td><td>"2019-04-08T11:01:20.000+02:00"</td></tr></table>|

Reviewer: Ce n'est pas une list

```

#### Update server

Sends a PATCH request  `/api/v2/servers/{SERVER_ID}` to update the informations of SERVER_ID.

```python
>>> CBWApi(URL, API_KEY, SECRET_KEY).update_server(SERVER_ID, INFO)
>>>True
```

#### Delete server

Sends a DELETE request `/api/v2/servers/{SERVER_ID}` to delete SERVER_ID.

```python
>>> CBWApi(URL, API_KEY, SECRET_KEY).delete_server(SERVER_ID)
>>>True
```

#### Remote accesses

Sends a GET request `/api/v2/remote_accesses` to retrieve the list of all remote accesses.

```python
>>> CBWApi(URL, API_KEY, SECRET_KEY).remote_accesses()
[<cbw_api_toolbox.cbw_objects.cbw_remote_access.CBWRemoteAccess...]
```

#### Remote access

Sends a GET request `/api/v2/remote_accesses/{REMOTE_ACCESS_ID}` to retrieve the informations of REMOTE_ACCESS_ID.

```python
>>> CBWApi(URL, API_KEY, SECRET_KEY).remote_access(REMOTE_ACCESS_ID)
[<cbw_api_toolbox.cbw_objects.cbw_remote_access.CBWRemoteAccess]
```

### Create remote access

Sends a POST request `/api/v2/remote_accesses` to create a remote access

```python
>>> CBWApi(URL, API_KEY, SECRET_KEY).create_remote_access(INFO)
INFO:root:remote access successfully created {ADDRESS}
True
```

### Update remote access

> Sends a PATCH request `/api/v2/remote_accesses/{REMOTE_ACCESS_ID}` to update the informations of REMOTE_ACCESSE_ID

Reviewer 2: ```suggestion
Sends a PATCH request `/api/v2/remote_accesses/{REMOTE_ACCESS_ID}` to update the informations of REMOTE_ACCESS_ID
```

```python
>>> CBWApi(URL, API_KEY, SECRET_KEY).update_remote_access(REMOTE_ACCESS_ID, INFO)
True
```

#### Delete remote access

Sends a DELETE request `/api/v2/remote_accesses/{REMOTE_ACCESS_ID}` to delete REMOTE_ACCESS_ID

```python
>>> CBWApi(URL, API_KEY, SECRET_KEY).delete_remote_access(REMOTE_ACCESS_ID)
True
```

## Available objects and their attributes

#### Server

| Attribute                 | Type          | Description                       | Exemple of possible value                         |
|---------------------------|:-------------:|:---------------------------------:|---------------------------------------------------|
| id                        | String (hash) | Unique id of the computer         | 4a78524087574c12453dea248a91cadb                  |

> | agent_version             | String        | Hostname of the computer          | DESKTOP-AAA111                                    |

Reviewer 2: agent_version is not the hostname

> | applications              | List          |                                   |                                                   |

Reviewer 2: Missing Description

| boot_at                   | Date          | Date of the computer boot         | 2016-05-30T09:54:58.000+02:00                     |

> | category                  | String        | Category of the machine           | server                                            |

Reviewer 2: ```suggestion
| category                  | String        | Computer category           | server                                            |
```

> | criticality               | String        | Criticality of the machine        | criticality_medium                                |

Reviewer 2: ```suggestion
| criticality               | String        | Computer criticality        | criticality_medium                                |
```

> | cve_announcements         | List          | All available listings            | CVE-2019-3842                                     |

Reviewer 2: What does "All available listings" mean?

> | cve_announcements_count   | Int           | Number of CVE                     | 3                                                 |    

Reviewer 2: ```suggestion
| cve_announcements_count   | Int           | Number of CVEs                     | 3                                                 |    
```

> | deploying_period          | List          | The time of deployment            | <table><tr><td>name:</td><td>"test"</td></tr><tr><td>autoplanning:</td><td>false</td></tr><tr><td>autoreboot:</td><td>false</td></tr><tr><td>start_time:</td><td>"00:00"</td><tr><td>end_time:</td><td>"06:00"</td></tr><tr><td>next_occurence:</td><td>"2019-04-27T00:00:00.000+02:00"</td></tr></table>|

Reviewer 2: s/The time of deployment/Computer deploying period

> | description               | String        | Description of the machine        | "machine of production"                           |

Reviewer 2: ```suggestion
| description               | String        | Computer description        | "production server"                           |
```

> | groups                    | List          | Groups of the machine             | "Production, Development"                         |

Reviewer 2: ```suggestion
| groups                    | List          | Computer groups             | "Production, Development"                         |
```

| hostname                  | String        | Name of the machine               | "XXX.XXX.XXX.XXX"                                 |

> | ignoring_policy           | List          |                                   | <table><tr><td>name:</td> <td>"test"</td></tr><tr><td>keyword:</td><td>"test"</td></tr><tr><td>version:</td><td>""</td></tr></table>|

Reviewer 2: Missing description

> | last_communication        | String        | Last communication of the machine | "2019-04-08T15:46:22.000+02:00"                   |

Reviewer 2: ```suggestion
| last_communication        | String        | Last communication of the computer | "2019-04-08T15:46:22.000+02:00"                   |
```

> | os                        | List          | Os of the machine                 | <table><tr><td>key:</td><td>"ubuntu_1804_64"</td></tr><tr><td>name:</td><td>"Ubuntu 18.04"</td></tr><tr><td>arch:</td><td>"x86_64"</td></tr><tr><td><eol:</td><td>"2023-04-26T02:00:00.000+02:00"</td></tr><tr><td>short_name:</td><td>"Ubuntu 18.04"</td></tr><tr><td>type:</td><td>"Os::Ubuntu"</td></tr><tr><td>hash_index:</td><td>"c3f4de66dd34cec0156bb439002bc4..."</td></tr><tr><td>created_at:</td><td>"2019-04-08T11:01:20.000+02:00"</td></tr><tr><td>updated_at:</td><td>"2019-04-08T11:01:20.000+02:00"</td></tr></table>|

Reviewer 2: OS of the computer

> | packages                  | List          |                                   | <table><tr><td>vendor:</td><td>null</td></tr><tr><td>product:</td><td>"gpg-wks-server"</td></tr><tr><td>hash_index:</td><td>"0160e4a3d38518a2660ff61b89c1407df"</td></tr><tr><td>type:</td><td>"Packages::Deb"</td></tr><tr><td>version:</td><td>"2.2.4-1ubuntu1.2"</td></tr></table>|

Reviewer 2: Missing description

> | reboot_required           | Bool          | obligation to restart the machine | false                                             |

Reviewer 2: ```suggestion
| reboot_required           | Bool          | The computer requires to be rebooted | false                                             |
```

> | remote_ip                 | String        | IP of the machine                 | XXX.XXX.XXX.XXX                                   |

Reviewer 2: ```suggestion
| remote_ip                 | String        | Computer IP address                | XXX.XXX.XXX.XXX                                   |
```

> | security_announcements    | List          |                                   | <table><tr><td>type:</td><td>"SecurityAnnouncements::Ubuntu"</td></tr><tr><td>sa_code:</td><td>"USN-3938-1"</td></tr><tr><td>created_at:</td><td>"2019-04-08T02:00:00.000+02:00"</td></tr><tr><td>updated_at:</td><td>"2019-04-10T09:20:07.000+02:00"</td></tr><tr><td>link:</td><td>"https://usn.ubuntu.com/3938-1"</td></tr><tr><td>cve_code:</td><td>"CVE-2019-3842"</td></tr></table>|

Reviewer 2: Missing description

> | state_sha2                | String        |                                   |                                                   |                                

Reviewer 2: What is state_sha2?

> | status                    | List          | Status of the rule on the computer| comment: "Communication failure"                  |

Reviewer 2: This is not the status of the rule, but just the status of the computer

```suggestion
| status                    | List          | Computer status | comment: "Communication failure"                  |
```

Reviewer 2: Also, the status_patching is missing

> | updates                   | List          | Update of the machine             | <table><tr><td>ignored:</td><td>false</td></tr><tr><td>patchable:</td><td>true</td></tr><th colspan="2">cve_announcements</th><tr><td>-id_cve:</td><td>22</td></tr><tr><td>-server_id:</td><td>21</td></tr><tr><td>-target_id:</td><td>239309</td></tr><tr><td>-current_id:</td><td>174477</td></tr><tr><td>-created_at:</td><td>"2019-04-09T17:13:07.000+02:00"</td></tr><tr><td>-updated_at:</td><td>"2019-04-09T17:13:07.000+02:00</td></tr><tr><td>-ignored:</td><td>false</td></tr><tr><td>-patchable:</td><td>true</td></tr><tr><th colspan="2">current</th></tr><tr><td>vendor:</td><td>null</td></tr><tr><td>product:</td><td>"libsystemd0"</td></tr><tr><td>hash_index:</td><td>"13a61dd2bfe0fbb063a97b66c6234759"</td></tr><tr><td>type:</td><td>"Packages::Deb"</td><tr><td>version:</td><td>"237-3ubuntu10.13"</td><tr><th colspan="2">target</th></tr><tr><td>vendor:</td><td>null</td></tr><tr><td>product:</td><td>"libsystemd0"</td></tr><tr><td>hash_index:</td><td>"58714c544b34b52d056475ee07ef41af"</td></tr><tr><td>type:</td><td>"Packages::Deb"</td></tr><tr><td>version:</td><td>"237-3ubuntu10.19"</td></tr><tr><th></th></tr><tr><td>created_at:</td><td>"2019-04-09T17:13:07.000+02:00"</td></tr><tr><td>updated_at:</td><td>"2019-04-09T17:13:07.000+02:00"</td></tr></table>|

Reviewer 2: ```suggestion
| updates                   | List          | Available updates for the computer             | <table><tr><td>ignored:</td><td>false</td></tr><tr><td>patchable:</td><td>true</td></tr><th colspan="2">cve_announcements</th><tr><td>-id_cve:</td><td>22</td></tr><tr><td>-server_id:</td><td>21</td></tr><tr><td>-target_id:</td><td>239309</td></tr><tr><td>-current_id:</td><td>174477</td></tr><tr><td>-created_at:</td><td>"2019-04-09T17:13:07.000+02:00"</td></tr><tr><td>-updated_at:</td><td>"2019-04-09T17:13:07.000+02:00</td></tr><tr><td>-ignored:</td><td>false</td></tr><tr><td>-patchable:</td><td>true</td></tr><tr><th colspan="2">current</th></tr><tr><td>vendor:</td><td>null</td></tr><tr><td>product:</td><td>"libsystemd0"</td></tr><tr><td>hash_index:</td><td>"13a61dd2bfe0fbb063a97b66c6234759"</td></tr><tr><td>type:</td><td>"Packages::Deb"</td><tr><td>version:</td><td>"237-3ubuntu10.13"</td><tr><th colspan="2">target</th></tr><tr><td>vendor:</td><td>null</td></tr><tr><td>product:</td><td>"libsystemd0"</td></tr><tr><td>hash_index:</td><td>"58714c544b34b52d056475ee07ef41af"</td></tr><tr><td>type:</td><td>"Packages::Deb"</td></tr><tr><td>version:</td><td>"237-3ubuntu10.19"</td></tr><tr><th></th></tr><tr><td>created_at:</td><td>"2019-04-09T17:13:07.000+02:00"</td></tr><tr><td>updated_at:</td><td>"2019-04-09T17:13:07.000+02:00"</td></tr></table>|
```

> | updates_count             | Int           | Number of update of the machine   | 3                                                 |

Reviewer 2: ```suggestion
| updates_count             | Int           | Number of available updates for the computer   | 3                                                 |
```
